File: Proj1/cpp/src/matrix.cpp
#include "matrix.hpp"

using namespace std;


Matrix::Matrix(size_t num_rows_, size_t num_cols_) {
    if(num_rows_ <= 0) {
        cerr << "[Wrong_Args] Number of rows must be greater than zero" << endl;
        exit(1);
    }

    if(num_cols_ <= 0){
        cerr << "[Wrong_Args] Number of columns must be greater than zero" << endl;
        exit(1);
    }

    num_rows = num_rows_;
    num_cols = num_cols_;

    matrix_values = new float[num_rows*num_cols];
}

Matrix::~Matrix() { delete[] matrix_values; }

void Matrix::fill(Matrix *matrix) {
    for(size_t i = 0; i < num_rows; i++) {
        for(size_t j = 0; j < num_cols; j++) {
            (*matrix)(i,j) = static_cast <float> (rand()) / static_cast <float> (101);
            //(*matrix)(i,j) = (rand() % 100 + 1); //random number between 1 and 100
        }
    }
}

void Matrix::fill_zeros(Matrix *matrix) {
    for(size_t i = 0; i < num_rows; i++) {
        for(size_t j = 0; j < num_cols; j++) {
            (*matrix)(i,j) = 0.0;
        }
    }
}

void Matrix::print(Matrix *matrix) {
    for(size_t i = 0; i < num_rows; i++) {
        for(size_t j = 0; j < num_cols; j++) {
            if(i==0 && j==0)
                cout << "[[" << (*matrix)(i,j) << ",";

            else if(j==0)
                cout << " [" << (*matrix)(i,j) << ",";

            else if(j==num_cols-1 && i==num_rows-1)
                cout << (*matrix)(i,j) << "]]";

            else if(j==num_cols-1)
                cout << (*matrix)(i,j) << "]";

            else
                cout << (*matrix)(i,j) << ",";
        }
        cout << endl;
    }

    cout << endl;
}

Matrix *Matrix::basic_multiply(Matrix *matrix_A, Matrix *matrix_B) {
    Matrix *result = new Matrix(matrix_A->num_rows, matrix_B->num_cols);
    fill_zeros(result);

    for(size_t i=0; i<matrix_A->num_rows; i++) {
        for(size_t j=0; j<matrix_B->num_cols; j++) {
            for(size_t k=0; k<matrix_A->num_cols; k++) {
                (*result)(i,j) += (*matrix_A)(i,k) * (*matrix_B)(k,j);
            }
        }
    }

    return result;
}

Matrix *Matrix::line_multiply(Matrix *matrix_A, Matrix *matrix_B) {
    Matrix *result = new Matrix(matrix_A->num_rows, matrix_B->num_cols);
    fill_zeros(result);

    for(size_t i=0; i<matrix_A->num_rows; i++) {
        for(size_t k=0; k<matrix_A->num_cols; k++) {
            for(size_t j=0; j<matrix_B->num_cols; j++) {
                (*result)(i,j) += (*matrix_A)(i,k) * (*matrix_B)(k,j);
            }
        }
    }

    return result;
}

Matrix *Matrix::omp_basic_multiply(Matrix *matrix_A, Matrix *matrix_B) {
    Matrix *result = new Matrix(matrix_A->num_rows, matrix_B->num_cols);
    size_t j, k;

    fill_zeros(result);
    #pragma omp parallel for private (j,k)
    for(size_t i=0; i<matrix_A->num_rows; i++) {
        for(j=0; j<matrix_B->num_cols; j++) {
            for(k=0; k<matrix_A->num_cols; k++) {
                (*result)(i,j) += (*matrix_A)(i,k) * (*matrix_B)(k,j);
            }
        }
    }

    return result;
}

Matrix *Matrix::omp_line_multiply(Matrix *matrix_A, Matrix *matrix_B) {
    Matrix *result = new Matrix(matrix_A->num_rows, matrix_B->num_cols);
    fill_zeros(result);
    size_t j, k;

    #pragma omp parallel for private (k,j)
    for(size_t i=0; i<matrix_A->num_rows; i++) {
        for(k=0; k<matrix_A->num_cols; k++) {
            for(j=0; j<matrix_B->num_cols; j++) {
                (*result)(i,j) += (*matrix_A)(i,k) * (*matrix_B)(k,j);
            }
        }
    }

    return result;
}

File: Proj1/py/src/matrix.py
from array import array

class Matrix:
    matrix_values = None
    num_rows = 0
    num_cols = 0

    def __init__(self, rows, columns):
        if rows <= 0:
            print '[Wrong_Args] Number of rows must be greater than zero'
            return 1

        if columns <= 0:
            print '[Wrong_Args] Number of columns must be greater than zero'
            return 1

        self.num_rows = rows
        self.num_cols = columns

        self.matrix_values = array("f", (0,) * (self.num_rows * self.num_cols))

def basic_multiply(matrix_A, matrix_B):
    result = Matrix(matrix_A.num_rows, matrix_B.num_cols)

    for i in range(matrix_A.num_rows):
        for j in range(matrix_B.num_cols):
            for k in range(matrix_A.num_cols):
                result.matrix_values[i*result.num_cols + j] += matrix_A.matrix_values[i*matrix_A.num_cols + k] * matrix_B.matrix_values[k*matrix_B.num_cols + j]

    return result

def line_multiply(matrix_A, matrix_B):
    result = Matrix(matrix_A.num_rows, matrix_B.num_cols)

    for i in range(matrix_A.num_rows):
        for k in range(matrix_A.num_cols):
            for j in range(matrix_B.num_cols):
                result.matrix_values[i*result.num_cols + j] += matrix_A.matrix_values[i*matrix_A.num_cols + k] * matrix_B.matrix_values[k*matrix_B.num_cols + j]

    return result
